File: PixivNovelHandler.py
import os
import time

from colorama.ansi import Fore, Style

import PixivConstant
import PixivHelper
import PixivNovel


def process_novel(caller,
                  config,
                  novel_id,
                  notifier=None):
    if notifier is None:
        notifier = PixivHelper.dummy_notifier

    msg = Fore.YELLOW + Style.BRIGHT + f'Processing Novel details: {novel_id}' + Style.RESET_ALL
    PixivHelper.print_and_log('info', msg)

    # check if already downloaded before and overwrite is not enabled
    db_result = caller.__dbManager__.selectNovelPostByPostId(novel_id)
    if db_result is not None and not config.overwrite and not config.checkLastModified and not config.alwaysCheckFileSize:
        save_name = db_result[2]  # save_name
        PixivHelper.print_and_log('warn', f"Novel already downloaded : {save_name}")
        return

    novel = caller.__br__.getNovelPage(novel_id)
    PixivHelper.print_and_log(None, f"Title : {novel.imageTitle}")
    PixivHelper.print_and_log(None, f'Member Name  : {novel.artist.artistName}')
    PixivHelper.print_and_log(None, f'Member Avatar: {novel.artist.artistAvatar}')
    PixivHelper.print_and_log(None, f'Member Token : {novel.artist.artistToken}')
    PixivHelper.print_and_log(None, f'Member Background : {novel.artist.artistBackground}')
    tags_str = ', '.join(novel.imageTags)
    PixivHelper.print_and_log(None, f"Tags : {tags_str}")
    PixivHelper.print_and_log(None, f"Date : {novel.worksDateDateTime}")
    PixivHelper.print_and_log(None, f"Mode : {novel.imageMode}")
    PixivHelper.print_and_log(None, f"Bookmark Count : {novel.bookmark_count}")

    # fake the fileUrl
    fileUrl = f"https://www.pixiv.net/ajax/novel/{novel_id}.html"
    filename = PixivHelper.make_filename(config.filenameFormatNovel,
                                         novel,
                                         tagsSeparator=config.tagsSeparator,
                                         tagsLimit=config.tagsLimit,
                                         fileUrl=fileUrl,
                                         bookmark=False,
                                         searchTags="",
                                         useTranslatedTag=config.useTranslatedTag,
                                         tagTranslationLocale=config.tagTranslationLocale)
    filename = PixivHelper.sanitize_filename(filename, config.rootDirectory)
    PixivHelper.print_and_log(None, f"Filename : {filename}")

    # checking logic
    if os.path.exists(filename):
        if config.checkLastModified:
            local_timestamp = os.path.getmtime(filename)
            remote_timestamp = time.mktime(novel.worksDateDateTime.timetuple())
            if local_timestamp == remote_timestamp:
                PixivHelper.print_and_log('warn', f"\rLocal file timestamp match with remote: {filename} => {novel.worksDateDateTime}")
                return
        if config.alwaysCheckFileSize:
            temp_filename = filename + ".!tmp"
            novel.write_content(temp_filename)
            file_size = os.path.getsize(temp_filename)
            old_size = os.path.getsize(filename)
            result = PixivHelper.check_file_exists(config.overwrite, filename, file_size, old_size, config.backupOldFile)
            if result == PixivConstant.PIXIVUTIL_OK:
                os.rename(temp_filename, filename)
            else:
                os.remove(temp_filename)
    else:
        novel.write_content(filename)

    if config.setLastModified and filename is not None and os.path.isfile(filename):
        ts = time.mktime(novel.worksDateDateTime.timetuple())
        os.utime(filename, (ts, ts))

    caller.__dbManager__.insertNovelPost(novel, filename)

    print()


def process_novel_series(caller,
                         config,
                         series_id,
                         start_page=1,
                         end_page=0,
                         notifier=None):
    if notifier is None:
        notifier = PixivHelper.dummy_notifier
    msg = Fore.YELLOW + Style.BRIGHT + f'Processing Novel Series: {series_id}' + Style.RESET_ALL
    PixivHelper.print_and_log('info', msg)

    novel_series = caller.__br__.getNovelSeries(series_id)
    PixivHelper.print_and_log(None, f'Series Name : {novel_series.series_name}')
    PixivHelper.print_and_log(None, f'Total Novel : {novel_series.total}')

    page = start_page
    flag = True
    while (flag):
        PixivHelper.print_and_log(None, f"Getting page = {page}")
        novel_series = caller.__br__.getNovelSeriesContent(novel_series, current_page=page)
        page = page + 1
        if end_page > 0 and page > end_page:
            PixivHelper.print_and_log(None, f"Page limit reached = {end_page}.")
            flag = False
        if (page * PixivNovel.MAX_LIMIT) >= novel_series.total:
            PixivHelper.print_and_log(None, "No more novel.")
            flag = False

    for novel in novel_series.series_list:
        process_novel(caller,
                      config,
                      novel_id=novel["id"],
                      notifier=notifier)

    print()

File: lib/handler/pixiv_novel_handler.dart
/// Novel handler.
library;

import 'dart:io';

import '../common/pixiv_browser.dart';
import '../common/pixiv_config.dart';
import '../common/pixiv_helper.dart' as pixiv_helper;
import '../model/pixiv_novel.dart' as pixiv_novel;

Future<void> processNovel({
  required dynamic caller,
  required PixivConfig config,
  required int novelId,
  void Function({String? title, String? message, dynamic type})? notifier,
}) async {
  notifier ??= pixiv_helper.dummyNotifier;
  pixiv_helper.printAndLog('info', 'Processing Novel details: $novelId');

  final br = caller.br as PixivBrowser;
  final novel = await br.getNovelPage(novelId);
  pixiv_helper.printAndLog(null, 'Title : ${novel.imageTitle}');
  pixiv_helper.printAndLog(null, 'Tags  : ${novel.imageTags?.join(', ')}');
  pixiv_helper.printAndLog(null, 'Date  : ${novel.worksDateDateTime}');
  pixiv_helper.printAndLog(
      null, 'Bookmark Count : ${novel.bookmark_count}');

  final fileUrl = 'https://www.pixiv.net/ajax/novel/$novelId.html';
  var filename = pixiv_helper.makeFilename(
    config.filenameFormatNovel,
    novel,
    artistInfo: novel.artist,
    tagsSeparator: config.tagsSeparator,
    tagsLimit: config.tagsLimit,
    fileUrl: fileUrl,
    useTranslatedTag: config.useTranslatedTag,
    tagTranslationLocale: config.tagTranslationLocale,
  );
  filename = pixiv_helper.sanitizeFilename(filename, config.rootDirectory);
  pixiv_helper.printAndLog(null, 'Filename : $filename');
  await novel.writeContent(filename);

  if (config.setLastModified) {
    try {
      File(filename).setLastModifiedSync(novel.worksDateDateTime);
    } catch (_) {}
  }
}

Future<void> processNovelSeries({
  required dynamic caller,
  required PixivConfig config,
  required int seriesId,
  int startPage = 1,
  int endPage = 0,
  void Function({String? title, String? message, dynamic type})? notifier,
}) async {
  notifier ??= pixiv_helper.dummyNotifier;
  pixiv_helper.printAndLog('info', 'Processing Novel Series: $seriesId');

  final br = caller.br as PixivBrowser;
  final series = await br.getNovelSeries(seriesId);
  pixiv_helper.printAndLog(null, 'Series Name : ${series.seriesName}');
  pixiv_helper.printAndLog(null, 'Total Novel : ${series.total}');

  var page = startPage;
  while (true) {
    pixiv_helper.printAndLog(null, 'Getting page = $page');
    final body = await br.getContent(
        'https://www.pixiv.net/ajax/novel/series_content/$seriesId?limit=${pixiv_novel.MAX_LIMIT}&last_order=${(page - 1) * pixiv_novel.MAX_LIMIT}');
    series.parseSeriesContent(body, page);
    if (endPage > 0 && page > endPage) {
      pixiv_helper.printAndLog(null, 'Page limit reached = $endPage');
      break;
    }
    if (page * pixiv_novel.MAX_LIMIT >= series.total) {
      pixiv_helper.printAndLog(null, 'No more novel.');
      break;
    }
    page++;
  }

  for (final novel in series.seriesList) {
    await processNovel(
      caller: caller,
      config: config,
      novelId: int.parse('${(novel as Map)['id']}'),
      notifier: notifier,
    );
  }
}
